src/index.js

```diff
@@ -83,18 +83,31 @@ function createListenForActions(store, stateFunc, actionListFunc, blacklistedAct
 
         resolve(stateFunc(store.getState()));
       } else {
-        unsubscribe = store.subscribe(() => {
-          // Get current action list
+        let callbackRan = false;
+
+        let resolver = () => {
           let actionListTypes = actionTypesFunc();
           if (_.isEqual(actionListTypes, expectedActionTypes)) {
             resolve(stateFunc(store.getState()));
           } else if (actionListTypes.length > expectedActionTypes.length) {
             reject(new Error("Received more actions than expected: actionListTypes: " +
               JSON.stringify(actionListTypes) + ", expectedActionTypes: " + JSON.stringify(expectedActionTypes)));
           }
+        };
+
+        unsubscribe = store.subscribe(() => {
+          // Get current action list
+          if (callbackRan) {
+            // wait until after callback() completes to resolve so we can detect thrown errors
+            // and actions which shouldn't have been dispatched
+            resolver();
+          }
         });
 
         callback();
+        callbackRan = true;
+        // If already resolved, this should get ignored
+        resolver();
```

Comment:
I can't pretend to fully understand everything going on here, but this logic looks straightforward enough - we subscribe to the store with a function that will run the resolver only after we set `callbackRan` to `true`, then we call the callback and set that boolean.

Before we were just setting that same subscription and then running the callback, which results in a race condition where if the expected actions complete before we hit an error in the callback we resolve the promise anyway and the error is swallowed.

is that accurate?